### PR TITLE
Bulk insert clean

### PR DESCRIPTION
**Pull request by AviAvni**

### CODE REVIEW

File: src/bulk_insert/bulk_insert.c [hunk 1]
```diff
@@ -150,70 +150,104 @@ static int _BulkInsert_ProcessFile(GraphContext *gc, const char *data,
...
 		}
 	}
 
+	Graph_SetMatrixPolicy(gc->g, SYNC_POLICY_RESIZE);
 	if(prop_indices) rm_free(prop_indices);
 	return BULK_OK;
 }
 
-static int _BulkInsert_ProcessTokens(GraphContext *gc, int token_count,
+static int _BulkInsert_ProcessNodeTokens(GraphContext *gc, int token_count,
```
Comment (swilly22) [inline src/bulk_insert/bulk_insert.c:230]: `_BulkInsert_ProcessEdgeTokens` and `_BulkInsert_ProcessNodeTokens` are practically the same consolidate
and use `type` to determine which function to call `_BulkInsert_ProcessNodeFile` or `_BulkInsert_ProcessEdgeFile`

File: src/bulk_insert/bulk_insert.c [hunk 2]
```diff
@@ -150,70 +150,104 @@ static int _BulkInsert_ProcessFile(GraphContext *gc, const char *data,
...
+		// process entity attributes
+		for(uint i = 0; i < prop_count; i++) {
+			SIValue value = _BulkInsert_ReadProperty(data, &data_idx);
+			// skip invalid attribute values
+			if(!(SI_TYPE(value) & SI_VALID_PROPERTY_VALUE)) continue;
+			GraphEntity_AddProperty(ge, prop_indices[i], value);
 		}
 	}
 
+	Graph_SetMatrixPolicy(gc->g, SYNC_POLICY_RESIZE);
```
Comment (swilly22) [inline src/bulk_insert/bulk_insert.c:225]: `SYNC_POLICY_RESIZE` is dangerous as it doesn't acquire matrix level locks
and bulk-insert can be run in parallel to queries.

**Review by swilly22 — CHANGES_REQUESTED** [review]

File: src/bulk_insert/bulk_insert.c (hunk 2 above)
Comment (AviAvni) [reply src/bulk_insert/bulk_insert.c:225]: this is the policy that was in line 294

Comment (AviAvni) [reply src/bulk_insert/bulk_insert.c:225]: also lock already acquired

File: src/bulk_insert/bulk_insert.c [hunk 3]
```diff
@@ -138,71 +138,91 @@ static SIValue _BulkInsert_ReadProperty(const char *data, size_t *data_idx) {
...
-	Attribute_ID *prop_indices = _BulkInsert_ReadHeader(gc, type, data,
+	Attribute_ID *prop_indices = _BulkInsert_ReadHeader(gc, SCHEMA_NODE, data,
 			&data_idx, &label_id, &prop_count);
 
-	if(type != SCHEMA_NODE && type != SCHEMA_EDGE) {
-		assert(false && "Bulk-Insert, unknown entity type");
-	}
+	
+	// sync matrix once
+	Graph_GetLabelMatrix(gc->g, label_id);
```
Comment (swilly22) [inline src/bulk_insert/bulk_insert.c:155]: `ASSERT(policy == resize)`

File: src/bulk_insert/bulk_insert.c [hunk 4]
```diff
@@ -138,71 +138,91 @@ static SIValue _BulkInsert_ReadProperty(const char *data, size_t *data_idx) {
...
+	int relation_id;
+	uint prop_count;
+	size_t data_idx = 0;
+
+	// read the CSV file header
+	// and commit all labels and properties it introduces
+	Attribute_ID *prop_indices = _BulkInsert_ReadHeader(gc, SCHEMA_EDGE, data,
+			&data_idx, &relation_id, &prop_count);
+
+	// sync matrix once
```
Comment (swilly22) [inline src/bulk_insert/bulk_insert.c:193]: ASSERT(policy == resize)

File: src/bulk_insert/bulk_insert.c [hunk 5]
```diff
@@ -213,7 +233,9 @@ static int _BulkInsert_ProcessTokens(GraphContext *gc, int token_count,
 		size_t len;
 		// retrieve a pointer to the next binary stream and record its length
 		const char *data = RedisModule_StringPtrLen(argv[i], &len);
-		int rc = _BulkInsert_ProcessFile(gc, data, len, type);
+		int rc = type == SCHEMA_NODE
```
Comment (swilly22) [inline src/bulk_insert/bulk_insert.c:236]:
```c
int rc = (type == SCHEMA_NODE)
```

File: src/bulk_insert/bulk_insert.c [hunk 6]
```diff
@@ -234,36 +256,34 @@ int BulkInsert(RedisModuleCtx *ctx, GraphContext *gc, RedisModuleString **argv,
...
 	}
 
+	Graph *g = gc->g;
+	int res = BULK_OK;
+
+	// lock graph under write lock
+	// allocate space for new nodes and edges
+	// set graph sync policy to resize only
+	Graph_SetMatrixPolicy(g, SYNC_POLICY_RESIZE);
+	Graph_AcquireWriteLock(g);
```
Comment (swilly22) [inline src/bulk_insert/bulk_insert.c:283]: swap, switch policy only after lock is acquired!

**Review by swilly22 — CHANGES_REQUESTED** [review]

**Review by swilly22 — APPROVED** [review]

**Pull request merged**